File: app/src/main/java/com/boardgamegeek/sorter/PlaysDateSorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Plays;
import com.boardgamegeek.util.DateTimeUtils;

import java.text.SimpleDateFormat;
import java.util.Locale;

import timber.log.Timber;

public class PlaysDateSorter extends PlaysSorter {
	private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("MMMM yyyy", Locale.getDefault());

	public PlaysDateSorter(@NonNull Context context) {
		super(context);
	}

	@StringRes
	@Override
	protected int getDescriptionId() {
		return R.string.menu_plays_sort_date;
	}

	@Override
	public int getType() {
		return PlaysSorterFactory.TYPE_PLAY_DATE;
	}

	@Override
	public String[] getColumns() {
		return new String[] { Plays.DATE };
	}

	@NonNull
	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		String date = getString(cursor, Plays.DATE);
		long millis = DateTimeUtils.getMillisFromApiDate(date, 0);
		if (millis == 0) {
			Timber.w("This isn't a date in the expected format: " + date);
			return date;
		}
		return DATE_FORMAT.format(millis);
	}
}

File: app/src/main/java/com/boardgamegeek/sorter/PlaysDateSorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import android.database.Cursor
import android.support.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.extensions.getApiTime
import com.boardgamegeek.provider.BggContract.Plays
import com.boardgamegeek.util.DateTimeUtils
import java.text.SimpleDateFormat
import java.util.*

class PlaysDateSorter(context: Context) : PlaysSorter(context) {
    private val defaultValue = context.getString(R.string.text_unknown)

    override val descriptionId: Int
        @StringRes
        get() = R.string.menu_plays_sort_date

    override val type: Int
        get() = PlaysSorterFactory.TYPE_PLAY_DATE

    override val columns: Array<String>
        get() = arrayOf(Plays.DATE)

    public override fun getHeaderText(cursor: Cursor): String {
        val time = cursor.getApiTime(Plays.DATE)
        return if (time == DateTimeUtils.UNKNOWN_DATE) defaultValue else DATE_FORMAT.format(time)
    }

    companion object {
        private val DATE_FORMAT = SimpleDateFormat("MMMM yyyy", Locale.getDefault())
    }
}
